contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,268 @@
+    update
+  - A DaemonSet pod needs to be ready for a specific amount of time before it's
+    considered up
+- Users can monitor the status of a DaemonSet update (e.g. the number of pods
+  that are updated and healthy)
+- A broken DaemonSet update should not continue, but one can still update the
+  DaemonSet again to fix it
+- Users should be able to update a DaemonSet even during an ongoing DaemonSet
+  upgrade -- in other words, rollover (e.g. update the DaemonSet to fix a broken
+  DaemonSet update)
+- Users should be able to view the history of previous DaemonSet updates 
+- Users can figure out the revision of a DaemonSet's pod (e.g. which version is
+  this DaemonSet pod?)
+
+Here are some potential requirements that haven't been covered by this proposal:
+
+- DaemonSet should provide at-most-one guarantee per node (i.e. at most one pod
+  from a DaemonSet can exist on a node at any time)
+- Uptime is critical for each pod of a DaemonSet during an upgrade (e.g. the time
+  from a DaemonSet pods being killed to recreated and healthy should be < 5s)
+- Each DaemonSet pod can still fit on the node after being updated
+- Some DaemonSets require the node to be drained before the DeamonSet's pod on it 
+  is updated (e.g. logging daemons)
+- DaemonSet's pods are implicitly given higher priority than non-daemons
+- DaemonSets can only be operated by admins (i.e. people who manage nodes)
+  - This is required if we allow DaemonSet controllers to drain, cordon,
+    uncordon nodes, evict pods, or allow DaemonSet pods to have higher priority
+
+## Implementation 
+
+### API Object 
+
+To enable DaemonSet upgrades, `DaemonSet` related API object will have the following
+changes:
+
+```go 
+type DaemonSetUpdateStrategy struct {
+	// Type of daemon set update. Can be "RollingUpdate" or "OnDelete". 
+	// Default is OnDelete.
+	Type DaemonSetUpdateStrategyType
+
+	// Rolling update config params. Present only if DaemonSetUpdateStrategy =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be. Same as DeploymentStrategy.RollingUpdate.
+	// See https://github.com/kubernetes/kubernetes/issues/35345
+	RollingUpdate *RollingUpdateDaemonSet
+}
+
+type DaemonSetUpdateStrategyType string
+
+const (
+	// Replace the old daemons by new ones using rolling update i.e replace them on each node one after the other.
+	RollingUpdateDaemonSetStrategyType DaemonSetUpdateStrategyType = "RollingUpdate"
+        
+	// Replace the old daemons only when it's killed
+	OnDeleteDaemonSetStrategyType DaemonSetUpdateStrategyType = "OnDelete"
+)
+
+// Spec to control the desired behavior of daemon set rolling update.
+type RollingUpdateDaemonSet struct {
+	// The maximum number of DaemonSet pods that can be unavailable during
+	// the update. Value can be an absolute number (ex: 5) or a percentage of total
+	// number of DaemonSet pods at the start of the update (ex: 10%). Absolute
```

Comment:
The total number will be calculated before killing daemon pods by counting the number of nodes that should run daemon pods. If nodes are added/removed during rollouts, `maxUnavailable` with a percentage may be violated.